src/runtime/micro/micro_common.h

```diff
@@ -0,0 +1,272 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file micro_common.h
+ */
+#ifndef TVM_RUNTIME_MICRO_MICRO_COMMON_H_
+#define TVM_RUNTIME_MICRO_MICRO_COMMON_H_
+
+#include <stdio.h>
+
+#include <tvm/runtime/registry.h>
+
+#include <sstream>
+#include <string>
+#include <unordered_map>
+
+namespace tvm {
+namespace runtime {
+/*!
+ * \brief enum of device memory region sections
+ */
+enum SectionKind : int {
+  kText = 0,
+  kRodata = 1,
+  kData = 2,
+  kBss = 3,
+  kArgs = 4,
+  kStack = 5,
+  kHeap = 6,
+  kWorkspace = 7,
+};
+
+// TODO(weberlo): There's a lot of duplication between these classes.  How can we consolidate?
+class dev_addr;
+class dev_base_addr;
+class dev_base_offset;
+
+/*! \brief absolute device address */
+class dev_addr {
```

Comment:
Consider using a base class with https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern to reduce the implementation duplication.

OpaquePtr